Multigroup radiation transport implementation

chongchonghe opened the pull request:
This PR implements multigroup radiation transport in Quokka. The key changes are:

- Radiation is now treated using N photon energy groups, with group boundaries and opacities defined for each group. 
- The radiation moment equations are solved separately for each group. Matter-radiation coupling is handled through a Newton-Raphson iteration that solves a set of 1+N equations with a sparse 1+N by 1+N Jacobian matrix. 
- Radiation energy exchange between gas and photons is apportioned among groups to conserve total energy. 
- Unit tests for the 1D RadForce and RadTube problems have been updated for the multigroup case. The RadTube test now verifies the separate evolution of two photon groups and compare it with analytic solution.
- ComputeRosselandOpacity is replaced with ComputeFluxMeanOpacity which returns the flux-mean opacity for each photon group. The energy-mean opacity is assumed to equal the flux-mean opacity. This assumption will be released via a future PR. 
- Infrastructure changes include adding the number of groups and group boundaries/units to traits classes, adding the number of variables for each fluid element, and refactoring some radiation-related data structures.

Additional quantitative test problems are planned for future PRs, such as a multigroup version of the SuOlson picket fence problem. This PR focuses on the core multigroup implementation and unit testing for 1D cases. The user should refer to the RadTube or RadForce example for usage of multigroup RT. For single-group (gray approximation) case, the only necessary change is to add `static constexpr int nGroups = 1;` to Physics_Traits. Feedback is welcome to further improve and test this new capability.

chongchonghe committed:
first commit: fix CMakeLists for macOS

chongchonghe committed:
multigroup branch, first try, with bugs

chongchonghe committed:
on macOS, at 2023-09-06 22:02:53

chongchonghe committed:
multigrou photon, second commit, fixed bug

Added multigroup photon support for radiation only.
The matter-radiation coupling is not added yet.

test_radiation_streaming: fully passed
test_radiation_force: produce the same result. The error caused by
ignoring matter-radiation coupling is negligible.
test_radiation_SuOlson: simulation runs but failed at logging stage.

chongchonghe committed:
Add radiation-matter coupling and radiation
energy update

- radiation_streaming is working
- SuOlson is kind of working

chongchonghe committed:
Multigropu RT, passed RadForce test

chongchonghe committed:
Multigroup RT, passed RadTube test

chongchonghe committed:
Multigroup: clean up

psharda commented on src/radiation_system.hpp:
```diff
@@ -28,10 +28,15 @@
 static constexpr double c_light_cgs_ = 2.99792458e10;	      // cgs
 static constexpr double radiation_constant_cgs_ = 7.5646e-15; // cgs
+static constexpr double pi = 3.14159265358979323846;
+static constexpr double hplanck = 6.62606957e-27;  // erg s
```
We don't need to define these again in quokka, we can just call the ones already defined in Microphysics. For example, k_B can be called as `C::k_B`

BenWibking replied:
Yes, good catch. We should change all of these to just use the values from Microphysics. They are all in the `C` namespace (C is for "constants").

BenWibking replied:
See this file for the available constants: https://github.com/AMReX-Astro/Microphysics/blob/main/constants/fundamental_constants.H

BenWibking commented:
@chongchonghe I can take a look at this in detail sometime later this week. In the meantime, can you fix the clang-tidy warnings? You can also see them locally in VSCode if you use the clangd extension (https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd)

See also: https://github.com/quokka-astro/quokka/wiki/How-to-use-clang%E2%80%90tidy

BenWibking commented on src/RadForce/test_radiation_force.cpp:
```diff
@@ -56,6 +56,11 @@ template <> struct RadSystem_Traits<TubeProblem> {
+	static constexpr double lower_rad_energy_bound = 0.0;
+	static constexpr double upper_rad_energy_bound = std::numeric_limits<double>::max();
+  static constexpr int nGroups = 2;
+  static constexpr std::array<double, nGroups - 1> radBoundaries {13.6};  // eV, size = nGroups - 1
```
I agree with Mark about changing this to `nGroups + 1` and including the lower energy bound and upper energy bound in this array.

BenWibking commented on src/RadForce/test_radiation_force.cpp:
```diff
@@ -67,14 +72,35 @@ template <> struct Physics_Traits<TubeProblem> {
 	// face-centred
 	static constexpr bool is_mhd_enabled = false;
+  // CCH: a hack to get nvarTotal_cc_radhydro right in physics_info.hpp
+  static constexpr int nGroups = RadSystem_Traits<TubeProblem>::nGroups;
```
I think it might be necessary to just define this in `Physics_Traits` and reference it within `RadSystem_Traits`, rather than defining it in both traits classes.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -133,18 +160,68 @@ template <typename problem_t> class RadSystem : public HyperbolicSystem<problem_
 
 	AMREX_GPU_HOST_DEVICE static auto ComputeEddingtonFactor(double f) -> double;
 	AMREX_GPU_HOST_DEVICE static auto ComputePlanckOpacity(double rho, double Tgas) -> double;
+	AMREX_GPU_HOST_DEVICE static auto ComputePlanckOpacityForGroup(double rho, double Tgas) -> quokka::valarray<double, nGroups_>;
```
I don't think it makes sense to have both `ComputePlanckOpacity` and `ComputePlanckOpacityForGroup`. I think it should be named `ComputePlanckOpacity` and it should just return an array of the opacities for all of the groups.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -133,18 +160,68 @@ template <typename problem_t> class RadSystem : public HyperbolicSystem<problem_
 	AMREX_GPU_HOST_DEVICE static auto ComputeEddingtonFactor(double f) -> double;
 	AMREX_GPU_HOST_DEVICE static auto ComputePlanckOpacity(double rho, double Tgas) -> double;
+	AMREX_GPU_HOST_DEVICE static auto ComputePlanckOpacityForGroup(double rho, double Tgas) -> quokka::valarray<double, nGroups_>;
+	AMREX_GPU_HOST_DEVICE static auto ComputeFluxMeanOpacityForGroup(double rho, double Tgas) -> quokka::valarray<double, nGroups_>;
```
Same comment as above.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -133,18 +160,68 @@ template <typename problem_t> class RadSystem : public HyperbolicSystem<problem_
+	AMREX_GPU_HOST_DEVICE static auto ComputePlanckOpacityForGroup(double rho, double Tgas) -> quokka::valarray<double, nGroups_>;
+	AMREX_GPU_HOST_DEVICE static auto ComputeFluxMeanOpacityForGroup(double rho, double Tgas) -> quokka::valarray<double, nGroups_>;
 	AMREX_GPU_HOST_DEVICE static auto ComputePlanckOpacityTempDerivative(double rho, double Tgas) -> double;
+	AMREX_GPU_HOST_DEVICE static auto ComputeOpacityTempDerivativeAtBins(double rho, double Tgas)->quokka::valarray<double, nGroups_>;
```
Same comment as above.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -133,18 +160,68 @@ template <typename problem_t> class RadSystem : public HyperbolicSystem<problem_
-	AMREX_GPU_DEVICE static auto ComputeCellOpticalDepth(const quokka::Array4View<const amrex::Real, DIR> &consVar,
-							     amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> dx, int i, int j, int k) -> double;
+  template <FluxDir DIR>
+  AMREX_GPU_DEVICE static auto ComputeCellOpticalDepth(const quokka::Array4View<const amrex::Real, DIR> &consVar,
```
This needs to be a function of the energy group.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -133,18 +160,68 @@ template <typename problem_t> class RadSystem : public HyperbolicSystem<problem_
 
+// CCH: compute radiation energy fractions for each bin of a Planck function, given the temperature.
+template <typename problem_t>
+void RadSystem<problem_t>::ComputePlanckEnergyFractions(quokka::valarray<amrex::Real, nGroups_> &radEnergyFractions, amrex::Real const temperature)
```
Is this needed for anything other than setting the initial conditions in LTE?

chongchonghe replied:
Yes, it's needed for calculating the source terms (Eq. 28 and 33 of the Quokka paper). Now, B is replace with B_i = B * fraction_i.

BenWibking commented on src/valarray.hpp:
```diff
@@ -59,6 +59,7 @@ template <typename T, int d> class valarray
 };
 } // namespace quokka
 
+// array + array
```
`valarray` is intended for representing mathematical vectors, with addition and multiplication by a scalar. I don't think it makes sense to use this object to add other operations. In general, it may be better to use a `std::array`.

chongchonghe replied:
I'm trying to extend it to the capacity of std::valarray. Is there anything wrong with this? I know the C++ community doesn't use valarray much (e.g. https://twitter.com/blelbach/status/839743137761259521?lang=en ), but I don't really understand why. I find it very convenient. Is it a speed issue?

BenWibking replied:
I agree that it is very convenient. I'm not sure why the C++ community doesn't like it very much.

If you want to extend it, I am okay with it. Could you put the changes you want to make to quokka::valarray in a separate PR so I can review them independently of the multigroup code? That PR can be reviewed and merged before tackling the this PR with all of the multigroup stuff.

BenWibking reviewed (commented):
I think the general approach to the code modifications is fine.

I think the additional variables you added to the traits classes can be simplified, so that RadSystem_Traits only needs one additional variable for the energy group boundaries and one additional variable to specify its units, and Physics_Traits only needs one additional variable for the number of energy groups.

I think you may be stretching `quokka::valarray` beyond its intended use case, which was originally for mathematical vectors with addition (and subtraction) of two vectors and multiplication by a scalar (and multiplication by the reciprocal of a scalar). In some cases, `std::array` may be a better choice. If that doesn't work, you may want to create a new array class for your purposes.

I don't think there should be multiple versions of the opacity functions. They should all return opacities for all of the energy groups. It doesn't make sense to keep the old opacity functions that return a scalar.

For multigroup test problems, I would suggest the multigroup radiative shock from Skinner et al. (https://arxiv.org/abs/1806.07390), the problem from Shestakov and Bolstad (https://ui.adsabs.harvard.edu/abs/2005JQSRT..91..133S/abstract), and the radiating sphere test from Graziani (2008) (https://link.springer.com/chapter/10.1007/978-3-540-77362-7_6).

@markkrumholz There is the more general question of whether to solve the comoving or mixed-frame equations. I think for multigroup the preferred option is actually to solve the comoving frame equations, because the group boundaries will otherwise change with the velocity boost. This is necessary to have a feasible number of energy groups where a small number of specific lines are of interest, e.g. Ly-alpha. But this can be implemented later.

chongchonghe committed:
tmp commit

chongchonghe committed:
some minor fixes

1. moved nGroups to Physics_Traits and put lower and upper bounds into radBoundaries
2. Renamed ComputePlanckOpacityForGroup to ComputePlanckOpacity.

chongchonghe committed:
renamed fillin in valarray

chongchonghe committed:
Implementing suggestions from Ben and Mark.
Two more major things to be done before ready to merge to
main branch after review:
1. update all tests to apply to multigroup (most of them are trivial)
2. update integrate_planck function following Mark's suggestion
- removing array_x.

BenWibking commented on src/RadForce/test_radiation_force.cpp:
```diff
@@ -67,14 +72,35 @@ template <> struct Physics_Traits<TubeProblem> {
 
-template <> AMREX_GPU_HOST_DEVICE auto RadSystem<TubeProblem>::ComputeRosselandOpacity(const double /*rho*/, const double /*Tgas*/) -> double { return kappa0; }
+template <> AMREX_GPU_HOST_DEVICE auto RadSystem<TubeProblem>::ComputePlanckOpacityForGroup(const double /*rho*/, const double /*Tgas*/) -> quokka::valarray<double, nGroups_>
+{
```
This is why this is getting a warning: https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/pro-type-member-init.html

BenWibking commented on src/RadForce/test_radiation_force.cpp:
```diff
@@ -67,14 +72,35 @@
+}
+
+template <> AMREX_GPU_HOST_DEVICE auto RadSystem<TubeProblem>::ComputeFluxMeanOpacityForGroup(const double /*rho*/, const double Tgas) -> quokka::valarray<double, nGroups_>
+{
```
Same issue: https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/pro-type-member-init.html

BenWibking commented:
All of the "uninitialized record type" warnings are due to this: https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/pro-type-member-init.html. Can you fix those?

chongchonghe committed:
define valarray.fillin and fixed radiation_tube

chongchonghe committed:
Ready to merge to main branch.
Passed 100% of the tests for 1D problems

chongchonghe commented:
Sorry. Still some work to do before ready for merging. Need to fix 2D and 3D tests. Also need to remove all the clang-tidy warnings.

chongchonghe committed:
Fixed compiling errors of the 2D and 3D problems
on multigroup branch

BenWibking commented:
It looks like the HIP build fails and the warnings CI fails. Can you take a look at those?

chongchonghe committed:
try to fix HIP errors and clang-tidy warnings

chongchonghe committed:
fix a few tidy warnings

chongchonghe committed:
fix ifndef bug

chongchonghe committed:
fix warning in SuOlson.cpp

chongchonghe committed:
fix a few tidy warnings

BenWibking commented:
/azp run

BenWibking commented:
The CUDA build produces these warnings:

```
../src/radiation_system.hpp(912): warning #780-D: reference is to structured binding "i"
(673): here -- under old for-init scoping rules it would have been variable "i"
(758): here

Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"

../src/radiation_system.hpp(913): warning #780-D: reference is to structured binding "i"
(673): here -- under old for-init scoping rules it would have been variable "i"
(758): here

../src/radiation_system.hpp(914): warning #780-D: reference is to structured binding "i"
(673): here -- under old for-init scoping rules it would have been variable "i"
(758): here

../src/radiation_system.hpp(915): warning #780-D: reference is to structured binding "i"
(673): here -- under old for-init scoping rules it would have been variable "i"
(758): here

../src/radiation_system.hpp(920): warning #780-D: reference is to structured binding "i"
(673): here -- under old for-init scoping rules it would have been variable "i"
(758): here

../src/radiation_system.hpp(921): warning #780-D: reference is to structured binding "i"
(673): here -- under old for-init scoping rules it would have been variable "i"
(758): here

../src/radiation_system.hpp(922): warning #780-D: reference is to structured binding "i"
(673): here -- under old for-init scoping rules it would have been variable "i"
(758): here

../src/radiation_system.hpp(923): warning #780-D: reference is to structured binding "i"
(673): here -- under old for-init scoping rules it would have been variable "i"
(758): here
```

Is this what you intended to do in this code?

BenWibking commented:
The CUDA build failed. Can you take a look?

chongchonghe committed:
fixed "radBoundaries not on device" error
for radiation_system.hpp and SuOlson test

chongchonghe committed:
redefined ComputePlanckEnergyFractions to return

chongchonghe committed:
cleanup tube.cpp

chongchonghe committed:
fix minor bugs

chongchonghe committed:
minor fix

chongchonghe commented:
/azp run

chongchonghe committed:
fixed "radBoundaries not on device" error.
Attempting to pass all tests on GPU.

chongchonghe committed:
/azp run

chongchonghe committed:
/azp run
Update valarray.cpp to match main branch

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chongchonghe committed:
fix 2 tidy warnings

chongchonghe committed:
Merge branch 'multigroup-dev0.1' of
https://github.com/quokka-astro/quokka
into multigroup-dev0.1

Github reformatted lots of the files without letting me know.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chongchonghe committed:
fix git conflicts

chongchonghe committed:
merge with remote

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chongchonghe committed:
fix 1 tidy warning

chongchonghe committed:
Merge branch 'multigroup-dev0.1' of https://github.com/quokka-astro/quokka into multigroup-dev0.1

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chongchonghe commented:
/azp run

BenWibking committed:
Merge branch 'development' into multigroup-dev0.1

BenWibking commented:
/azp run

BenWibking commented:
I'll try to review this in detail over the weekend.

Aside from the code itself, I am hesitant to approve this without quantative tests, although they could go in future PRs. What is the plan for quantitative test problems that check the group-by-group solution?

chongchonghe committed:
minor tidy fix

chongchonghe committed:
Merge branch 'multigroup-dev0.1' of https://github.com/quokka-astro/quokka into multigroup-dev0.1

BenWibking commented:
/azp run

BenWibking committed:
Merge branch 'development' into multigroup-dev0.1

BenWibking commented:
/azp run

BenWibking commented on src/RadForce/test_radiation_force.cpp:
```diff
@@ -66,14 +59,42 @@ template <> struct Physics_Traits<TubeProblem> {
+    -> quokka::valarray<double, Physics_Traits<TubeProblem>::nGroups>
+{
+	quokka::valarray<double, Physics_Traits<TubeProblem>::nGroups> kappaFVec{};
+	// kappaFVec.fillin(kappa0);
```
Can you remove commented-out code? Unless there's a very good reason to keep it, it just makes it less readable to future developers/users. We've tried to avoid keeping commented-out code anywhere in Quokka (although there are a few places where it still needs to be removed).

chongchonghe replied:
I removed it and all other similar things.

BenWibking commented on src/RadForce/test_radiation_force.cpp:
```diff
@@ -132,6 +153,13 @@ template <> void RadhydroSimulation<TubeProblem>::setInitialConditionsOnGrid(quo
 
+	// CCH: calculate radEnergyFractions
+	quokka::valarray<amrex::Real, Physics_Traits<TubeProblem>::nGroups> radEnergyFractions{};
+	// radEnergyFractions.fillin(1.0 / Physics_Traits<TubeProblem>::nGroups);
```
Can you remove the commented-out code?

BenWibking commented on src/RadForce/test_radiation_force.cpp:
```diff
@@ -132,6 +153,13 @@ template <> void RadhydroSimulation<TubeProblem>::setInitialConditionsOnGrid(quo
 	auto const &Mach_ptr = Mach_arr_g.dataPtr();
 	int x_size = static_cast<int>(x_arr_g.size());
 
+	// CCH: calculate radEnergyFractions
```
As a matter of style, we haven't added initials indicating the developer who has written the code, since it's always available with `git blame`.

chongchonghe replied:
All resolved.

BenWibking commented on src/RadForce/test_radiation_force.cpp:
```diff
@@ -183,6 +215,15 @@ AMRSimulation<TubeProblem>::setCustomBoundaryConditions(const amrex::IntVect &iv
 
+	// CCH: calculate radEnergyFractions
+	// amrex::Real const temperature = 0.0;
+	// RadSystem<TubeProblem>::ComputeRadEnergyFractions(radEnergyFractions, temperature);
```
Remove the commented-out code.

chongchonghe replied:
Done

BenWibking commented on src/RadForce/test_radiation_force.cpp:
```diff
@@ -183,6 +215,15 @@ AMRSimulation<TubeProblem>::setCustomBoundaryConditions(const amrex::IntVect &iv
+	// amrex::Real const temperature = 0.0;
+	// RadSystem<TubeProblem>::ComputeRadEnergyFractions(radEnergyFractions, temperature);
+	quokka::valarray<amrex::Real, Physics_Traits<TubeProblem>::nGroups> radEnergyFractions{};
+	// radEnergyFractions.fillin(1.0 / Physics_Traits<TubeProblem>::nGroups);
```
Remove the commented-out code.

BenWibking commented on src/RadForce/test_radiation_force.cpp:
```diff
@@ -183,6 +215,15 @@ AMRSimulation<TubeProblem>::setCustomBoundaryConditions(const amrex::IntVect &iv
 	amrex::Real rho = NAN;
 	amrex::Real vel = NAN;
 
+	// CCH: calculate radEnergyFractions
```
This is not strictly necessary, since future users can always find that you wrote this line of code with `git blame`.

chongchonghe replied:
Resolved this and all similar comments.

BenWibking commented on src/RadMarshak/test_radiation_marshak.cpp:
```diff
@@ -52,13 +52,24 @@ template <> struct Physics_Traits<SuOlsonProblem> {
+AMREX_GPU_HOST_DEVICE auto RadSystem<SuOlsonProblem>::ComputePlanckOpacity(const double /*rho*/, const double /*Tgas*/) -> quokka::valarray<double, nGroups_>
+{
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	// kappaPVec.fillin(kappa);
```
Remove commented-out code.

BenWibking commented on src/RadMarshakAsymptotic/test_radiation_marshak_asymptotic.cpp:
```diff
@@ -47,25 +47,40 @@ template <> struct Physics_Traits<SuOlsonProblemCgs> {
 	auto sigma = kappa * std::pow(Tgas / T_hohlraum, -3); // cm^-1
-	return (sigma / rho);				      // cm^2 g^-1
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	// kappaPVec.fillin(sigma / rho);      // cm^2 g^-1
```
Remove commented-out code.

BenWibking commented on src/RadMarshakAsymptotic/test_radiation_marshak_asymptotic.cpp:
```diff
@@ -47,25 +47,40 @@ template <> struct Physics_Traits<SuOlsonProblemCgs> {
+	quokka::valarray<double, nGroups_> opacity_deriv{};
 	auto sigma_dT = (-3.0 * kappa / Tgas) * std::pow(Tgas / T_hohlraum, -3); // cm^-1
-	return (sigma_dT / rho);
+	// opacity_deriv.fillin(sigma_dT / rho);
```
Remove commented-out code.

BenWibking commented on src/RadMarshakCGS/test_radiation_marshak_cgs.cpp:
```diff
@@ -54,16 +54,25 @@ template <> struct Physics_Traits<SuOlsonProblemCgs> {
 {
-	return kappa;
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	// kappaPVec.fillin(kappa);
```
Remove commented-out code.

BenWibking commented on src/RadMatterCoupling/test_radiation_matter_coupling.cpp:
```diff
@@ -52,13 +52,24 @@ template <> struct Physics_Traits<CouplingProblem> {
+AMREX_GPU_HOST_DEVICE auto RadSystem<CouplingProblem>::ComputePlanckOpacity(const double /*rho*/, const double /*Tgas*/) -> quokka::valarray<double, nGroups_>
+{
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	// kappaPVec.fillin(1.0);
```
Remove commented-out code.

BenWibking commented on src/RadMatterCouplingRSLA/test_radiation_matter_coupling_rsla.cpp:
```diff
@@ -54,13 +54,24 @@ template <> struct Physics_Traits<CouplingProblem> {
+AMREX_GPU_HOST_DEVICE auto RadSystem<CouplingProblem>::ComputePlanckOpacity(const double /*rho*/, const double /*Tgas*/) -> quokka::valarray<double, nGroups_>
+{
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	// kappaPVec.fillin(1.0);
```
Remove commented-out code.

BenWibking commented on src/RadStreaming/test_radiation_streaming.cpp:
```diff
@@ -44,16 +37,33 @@ template <> struct Physics_Traits<StreamingProblem> {
 {
-	return kappa0;
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	// kappaPVec.fillin(kappa0);
```
Remove commented-out code.

BenWibking commented on src/RadStreaming/test_radiation_streaming.cpp:
```diff
@@ -64,12 +74,22 @@ template <> void RadhydroSimulation<StreamingProblem>::setInitialConditionsOnGri
 
+	// CCH: calculate radEnergyFractions
+	quokka::valarray<amrex::Real, Physics_Traits<StreamingProblem>::nGroups> radEnergyFractions{};
+	// radEnergyFractions.fillin(1.0 / Physics_Traits<StreamingProblem>::nGroups);
```
Remove commented-out code.

BenWibking commented on src/RadStreaming/test_radiation_streaming.cpp:
```diff
@@ -102,23 +122,36 @@ AMRSimulation<StreamingProblem>::setCustomBoundaryConditions(const amrex::IntVec
 
+	// CCH: calculate radEnergyFractions
+	quokka::valarray<amrex::Real, Physics_Traits<StreamingProblem>::nGroups> radEnergyFractions{};
+	// radEnergyFractions.fillin(1.0 / Physics_Traits<StreamingProblem>::nGroups);
```
Remove commented-out code.

BenWibking commented on src/RadSuOlson/test_radiation_SuOlson.cpp:
This test problem is designed for gray radiation transport. Does it make sense to make this a 2-group problem?

chongchonghe replied:
You are right. I changed it back to 1 group. In a seperate PR, I'll add the two-gropu SuOlson with Picket-Fence test following Sec 4.1.2 of the CASTRO paper.

BenWibking commented on src/RadTube/test_radiation_tube.cpp:
```diff
@@ -58,11 +52,36 @@ template <> struct Physics_Traits<TubeProblem> {
+{
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	// amrex::Real const kappa0_g = kappa0;
+	// kappaPVec.fillin(kappa0_g);
```
Remove commented-out code.

BenWibking commented on src/RadhydroShell/test_radhydro_shell.cpp:
```diff
@@ -123,15 +124,21 @@ void RadSystem<ShellProblem>::SetRadEnergySource(array_t &radEnergy, const amrex
 {
-	return kappa0;
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	// kappaPVec.fillin(kappa0);
```
Remove commented-out code.

BenWibking commented on src/RadhydroShock/test_radhydro_shock.cpp:
```diff
@@ -75,16 +75,25 @@ template <> struct Physics_Traits<ShockProblem> {
 {
-	return (kappa / rho);
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	// kappaPVec.fillin(kappa / rho);
```
Remove commented-out code.

BenWibking commented on src/RadhydroShockCGS/test_radhydro_shock_cgs.cpp:
```diff
@@ -76,16 +76,25 @@ template <> struct Physics_Traits<ShockProblem> {
 {
-	return (kappa / rho);
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	// kappaPVec.fillin(kappa / rho);
```
Remove commented-out code.

BenWibking commented on src/RadhydroSimulation.hpp:
```diff
@@ -297,7 +297,14 @@ template <typename problem_t> void RadhydroSimulation<problem_t>::defineComponen
 	// add radiation state variables
 	if constexpr (Physics_Traits<problem_t>::is_radiation_enabled) {
-		std::vector<std::string> radNames = {"radEnergy", "x-RadFlux", "y-RadFlux", "z-RadFlux"};
+		// std::vector<std::string> radNames = {"radEnergy", "x-RadFlux", "y-RadFlux", "z-RadFlux"};
```
Remove commented-out code.

BenWibking commented on src/RadhydroSimulation.hpp:
```diff
@@ -543,9 +550,17 @@ template <typename problem_t> void RadhydroSimulation<problem_t>::computeAfterEv
 	amrex::Real Etot = NAN;
 	if constexpr (Physics_Traits<problem_t>::is_radiation_enabled) {
-		amrex::Real const Erad0 = initSumCons[RadSystem<problem_t>::radEnergy_index];
+		// amrex::Real const Erad0 = initSumCons[RadSystem<problem_t>::radEnergy_index];
```
Remove commented-out code.

BenWibking commented on src/RadhydroSimulation.hpp:
```diff
@@ -543,9 +550,17 @@ template <typename problem_t> void RadhydroSimulation<problem_t>::computeAfterEv
+		}
 		Etot0 = Egas0 + (RadSystem<problem_t>::c_light_ / RadSystem<problem_t>::c_hat_) * Erad0;
-		amrex::Real const Erad = state_new_cc_[0].sum(RadSystem<problem_t>::radEnergy_index) * vol;
+		// amrex::Real const Erad = state_new_cc_[0].sum(RadSystem<problem_t>::radEnergy_index) * vol;
```
Remove commented-out code.

BenWibking commented on tests/Backtrace.0:
This file should be removed from the PR.

BenWibking commented on src/planck_integral.hpp:
```diff
@@ -0,0 +1,241 @@
+
+using Real = amrex::Real;
+
+static constexpr double PI = 3.14159265358979323846;
```
Although it is not part of the standard, can you use `M_PI` instead (https://www.gnu.org/software/libc/manual/html_node/Mathematical-Constants.html)?

BenWibking replied:
Once we switch to C++20, we can use `std::numbers::pi`: https://en.cppreference.com/w/cpp/numeric/constants

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -26,12 +27,16 @@
-static constexpr double radiation_constant_cgs_ = 7.5646e-15; // cgs
+static constexpr double c_light_cgs_ = C::c_light;	    // cgs
+static constexpr double radiation_constant_cgs_ = C::a_rad; // cgs
+static constexpr double pi = 3.14159265358979323846;
```
Can you use `M_PI` instead?

chongchonghe replied:
Done.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -132,19 +150,76 @@ template <typename problem_t> class RadSystem : public HyperbolicSystem<problem_
+{
+	quokka::valarray<amrex::Real, nGroups_> radEnergyFractions{};
+	if constexpr (nGroups_ == 1) {
+		// CCH: for future reference, if we want to use a single group as a delta function, say at 13.6 eV, rethink this.
```
As a matter of code style, you could write this as `// TODO(cch): for future reference...`

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -132,19 +150,76 @@ template <typename problem_t> class RadSystem : public HyperbolicSystem<problem_
+		// TODO(CCH): find a better way to do this
+		for (int g = 0; g < nGroups_; ++g) {
+			if (radEnergyFractions[g] < 1e-14) {
+				radEnergyFractions[g] = 1e-14;
```
Can you change this to be a variable instead of a hard-coded value of `1e-14`? A default value for this variable of 1e-14 seems fine to me, but it should not be hard-coded if we can avoid it.

chongchonghe replied:
Done.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -611,264 +691,284 @@ void RadSystem<problem_t>::ComputeFluxes(array_t &x1Flux_in, array_t &x1FluxDiff
 }
 
-template <typename problem_t> AMREX_GPU_HOST_DEVICE auto RadSystem<problem_t>::ComputeRosselandOpacity(const double /*rho*/, const double /*Tgas*/) -> double
+// TODO(CCH): Derivative of Planck/energy-mean opacity with respect to gas temperature. In the future, they should be defined separately.
```
I don't understand what you mean here. Can you explain this comment?

chongchonghe replied:
Similar to my other comments about d B_i / d T_gas, this is something I plan to do next. Mark and I discussed about this, and our plan is to define a class to allow the user to specify d kappa_P / d Tgas. We will also provide a convenient function to compute this from the Integral of Planck function if the user provide a tabulated kappa(nu). I removed this TODO and kept the other one about d B_i / d T_gas. 

I removed all TODOs except a few of them which I find it necessary to keep.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
+		}
+
+		// if constexpr (gamma_ != 1.0) {
+		if (gamma_ != 1.0) {
```
Does this not work with `constexpr` anymore?

chongchonghe replied:
`constexpr` works now. I've added it back.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
+				auto B = fourPiB / (4.0 * pi);
+				kappaPVec = ComputePlanckOpacity(rho, T_gas);
+				kappaFVec = ComputeFluxMeanOpacity(rho, T_gas);
+				kappaEVec = kappaFVec; // TODO(CCH): later, define ComputeEnergyMeanOpacity
```
Can you explain what you mean here?

chongchonghe replied:
There are kappa_P, kappa_F, and kappa_E. Currently I'm assuming kappa_E = kappa_F. In the future we want to release this assumption.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
+				// prepare to compute Jacobian elements
+				const double c_v = quokka::EOS<problem_t>::ComputeEintTempDerivative(rho, T_gas, massScalars);
+				dRtot_dErad = -dt * rho * kappaEVec * chat;
+				// TODO(CCH): fix this: dB_dTgas * radEnergyFractions is inaccurate. Define ComputePlanckEnergyFractionsTempDerivative
```
Can you explain this?

chongchonghe replied:
To calculate d B_i / d Tgas accurately, I need to consider that both B_i and radFractions_i is a function of Tgas. Currently I am using an approximation that d B_i / d Tgas = radFractions_i * d B / d Tgas, which is exact in 1-group case and mostly accurate in multigropu case. The error caused by this is below 0.2% in the RadTube test. I'll do it rigorously in the next PR.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
+					Frad_t1[2] = consNew(i, j, k, x3RadFlux_index + numRadVars_ * g);
+					energyLossFractions[g] = kappaFVec[g] * (x1GasMom1 * Frad_t1[0] + x2GasMom1 * Frad_t1[1] + x3GasMom1 * Frad_t1[2]);
+				}
+				// TODO(CCH): tricky here. Have to avoid negavie radiation energy.
```
Can you explain the negative energy problem you encountered?

chongchonghe replied:
Never mind this. It shouldn't be a TODO and I removed it.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
 	});
 }
 
+// CCH: this function ComputeSourceTermsExplicit is never used.
```
Yes, we can get rid of it in another PR.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
 			}
 
-			// compute difference between new and old internal energy
-			amrex::Real const dEint = Egas_guess - Egas0;
```
Is there a reason to get rid of this variable?

chongchonghe replied:
Not really. I like to keep the code concise and avoid unnecessary variables, but I also appreciate extra variables to better explain the physics. I put dEint back.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
+				dRtot_dEgas = sum(dRvec_dEgas);
+				// kappaP = RadSystem<problem_t>::ComputePlanckOpacity(rho, T_gas);
+				// dRtot_dEgas = dt * rho / c_v * (4 * pi * kappaP * dB_dTgas + dkappa_dTgas * fourPiB - chat * sum(dkappaVec_dTgas *
+				// EradVec_guess));
```
Can you remove the commented-out code?

chongchonghe replied:
Done.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
+				// TODO(CCH): fix this: if constexpr (nGroups_ > 1)
+				// if constexpr (nGroups_ > 1) {
+				// ../src/radiation_system.hpp(1089): error: An extended __device__ lambda cannot first-capture variable in constexpr-if context
+				// detected during: ...
```
This is an annoying limitation of the CUDA compiler. You can use this variable before the `ParallelFor` and then it will work.

chongchonghe replied:
Good. Actually I realised this before but forgot to put the constexpr back.

BenWibking reviewed (changes requested):
There are a few style changes. In particular, we have tried to avoid commented-out code in Quokka in general. Unless there is a very good reason to keep it, it is usually available in the git history, or it never worked and was never part of the code, in which case it just confuses new developers/users.  We have also avoiding adding initials or developer names in comments, since this information is always available via `git blame`, although it's up to you if you want to keep that.

The main thing I am concerned about are the partially-explained comments in the radiation-matter update code that indicate TODO items. Can you explain those further? I think it's important to know if there are issues with the current implementation, or if you are just keeping track of where future improvements could be made.

chongchonghe committed:
removed some commented code

chongchonghe committed:
Removed comments with author initials;
removed unnecessary TODOs

chongchonghe commented:
> I'll try to review this in detail over the weekend.
> 
> Aside from the code itself, I am hesitant to approve this without quantative tests, although they could go in future PRs. What is the plan for quantitative test problems that check the group-by-group solution?

Let's discuss this on our next Quokka meeting.

BenWibking commented:
@chongchonghe @markkrumholz For mixed-frame, do the group boundaries need to be Doppler-shifted before doing the radiation-matter energy exchange? I think they do, but maybe I'm wrong.

BenWibking commented:
> > I'll try to review this in detail over the weekend.
> > Aside from the code itself, I am hesitant to approve this without quantative tests, although they could go in future PRs. What is the plan for quantitative test problems that check the group-by-group solution?
> 
> Let's discuss this on our next Quokka meeting.

For reference, see section 4.1: https://arxiv.org/abs/1207.3845.

markkrumholz commented:
Just a few quick comments on the physics here. First, we do not need to Doppler shift the group boundaries. In the mixed-frame formulation, all frequencies are written on the lab frame, and I do think we want to keep using mixed-frame, for reasons I’ll get to in a moment. 

Second, there are terms that involve shifts between energy bins due to work being done on the photons by matter, which we have not yet implemented. These are small for our problems, but should be included. I am unsure if it is with going the CASTRO route of treating them as an advection, or whether we should group them with the source terms. CASTRO does the former, but they’re doing problems with v/c ~ 0.1, whereas we are at ~10^-3. 

Third, I think we want to stick with mixed-frame because we are not doing line transfer. The only sharp features we care about in our opacities are ionization edges, and the mixed-frame formulation does fail near them. However, this is a tiny part of our frequency bin, so the error is small. Quantitatively, if for example we have a bin that goes from E_0 = 13.6 to E_1 = 20 eV, so Delta E = 6.4 eV then the error we make by applying mixed-frame is of order (v/c) E_0 / Delta E ~ 10^-3. This is smaller than the typical conservation error you would get in a comoving frame formulation.

BenWibking commented:
> Third, I think we want to stick with mixed-frame because we are not doing line transfer. The only sharp features we care about in our opacities are ionization edges, and the mixed-frame formulation does fail near them. However, this is a tiny part of our frequency bin, so the error is small. Quantitatively, if for example we have a bin that goes from E_0 = 13.6 to E_1 = 20 eV, so Delta E = 6.4 eV then the error we make by applying mixed-frame is of order (v/c) E_0 / Delta E ~ 10^-3. This is smaller than the typical conservation error you would get in a comoving frame formulation.

I would like to do line transfer in the future. However, maybe we can just add a runtime check that the bins are sufficiently coarse near ionization edges in the current implementation.

chongchonghe committed:
add numeric check of E_1 and E_2 in SuOlson

BenWibking commented:
@chongchonghe Can you take a look at the remaining unresolved comments?

BenWibking committed:
Merge branch 'development' into multigroup-dev0.1

chongchonghe committed:
Merge branch 'multigroup-dev0.1' of https://github.com/quokka-astro/quokka into multigroup-dev0.1

chongchonghe committed:
add quadratic interp support to planck_integral

chongchonghe committed:
Add E_0 and E_1 to L1 norm check in RadTube test.

chongchonghe committed:
reset SuOlson to 1gropu;
removed commented-out code.

chongchonghe commented:
> There are a few style changes. In particular, we have tried to avoid commented-out code in Quokka in general. Unless there is a very good reason to keep it, it is usually available in the git history, or it never worked and was never part of the code, in which case it just confuses new developers/users. We have also avoiding adding initials or developer names in comments, since this information is always available via `git blame`, although it's up to you if you want to keep that.
> 
> The main thing I am concerned about are the partially-explained comments in the radiation-matter update code that indicate TODO items. Can you explain those further? I think it's important to know if there are issues with the current implementation, or if you are just keeping track of where future improvements could be made.

I removed all unnecessary TODOs and kept only a few necessary ones. They are not issues. They are planned work that will release some assumptions/approximates.

chongchonghe committed:
Improved styling. Cleaned up comments.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chongchonghe commented:
> I'll try to review this in detail over the weekend.
> 
> Aside from the code itself, I am hesitant to approve this without quantative tests, although they could go in future PRs. What is the plan for quantitative test problems that check the group-by-group solution?

There is one quantitative test in the current branch, the RadTube test. I set the boundary between two photon groups at near the temperature of the tube. In the new figure, radiation_pressure_tube_energy_density.pdf, you'll see that the exact solution for Erad_1 and Erad_2 are on top of the numerical curve. The proportions of the two groups changes with the temperature which changes from left to right. That's why you see it changed from group1-dominating to group2-dominating from left to right. The exact solution is calculated from the initial/equilibrium solution of Erad and Tgas and stored in the cpp file of the problem. The proportions are calculated using a Python script. 

In a seperate PR, I'll add the multigroup SuOlson picket-fence test (Su & Olson 1999, Case C, see also http://arxiv.org/abs/1207.3845). I may also add a few other tests, but they are not planed.

chongchonghe committed:
Merge branch 'development' into multigroup-dev0.1

chongchonghe committed:
fixed variable unused error

chongchonghe committed:
merge development in

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chongchonghe committed:
fix warning

chongchonghe committed:
Merge branch 'multigroup-dev0.1' of https://github.com/quokka-astro/quokka into multigroup-dev0.1

BenWibking commented:
/azp run

BenWibking commented:
The GPU build failed due to the first-capture issue: https://dev.azure.com/quokka-code/quokka-CI/_build/results?buildId=6772&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=d735d76d-5905-5d5d-cb1c-7198326fbfa2

BenWibking committed:
Merge branch 'development' into multigroup-dev0.1

BenWibking commented:
Could you add a somewhat more detailed PR description? Since we don't have real documentation, our best existing documentation has been PR descriptions. About 1 paragraph in length would be good, along with a description of any test problems you've added, and any limitations or assumptions that potential users of the multigroup code should be aware of.

Once this is done and the CUDA build works, I'm happy to approve and merge.

chongchonghe committed:
fixed first-capture issue

chongchonghe committed:
Merge branch 'multigroup-dev0.1' of https://github.com/quokka-astro/quokka into multigroup-dev0.1

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chongchonghe commented:
/azp run

BenWibking commented on extern/Microphysics:
It looks like your submodules are out of date. Can you update your submodules?

chongchonghe replied:
I don't see any outdated submodules. I believe they are up to date.

chongchonghe commented:
/azp run

chongchonghe committed:
update submodule Microphysics

chongchonghe committed:
update submodule amrex

chongchonghe committed:
update submodule fmt

chongchonghe committed:
update submodule yaml-cpp

chongchonghe committed:
update submodule amrex to correct branch

chongchonghe committed:
update fmt to correct branch

BenWibking commented:
/azp run

BenWibking commented:
Can you update the PR description?

BenWibking commented on src/RadPulse/test_radiation_pulse.cpp:
```diff
@@ -63,22 +65,33 @@ auto compute_exact_Trad(const double x, const double t) -> double
+		opacity_deriv_scalar = (kappa0 / rho) * (3.0 / T0) * std::pow(Tgas / T0, 2);
 	}
-	return 0.;
+	opacity_deriv.fillin(opacity_deriv_scalar);
```
Does `fillin` work on GPU?

chongchonghe replied:
It works on variables that are defined locally or hand-written number like 0.0. It doesn't work on globally defined constexpr like kappa0.

BenWibking commented on src/radiation_system.hpp:
```diff
@@ -611,264 +689,280 @@ void RadSystem<problem_t>::ComputeFluxes(array_t &x1Flux_in, array_t &x1FluxDiff
 {
-	return NAN;
+	quokka::valarray<double, nGroups_> kappa{};
+	kappa.fillin(0.0);
```
Does `fillin` work on GPU?

chongchonghe commented:
> Can you update the PR description?

I'll do it this noon when I get to work. What is exactly a PR description? Is it the first comments I made in the PR? I just edit that?

BenWibking commented:
> > Can you update the PR description?
> 
> I'll do it this noon when I get to work. What is exactly a PR description? Is it the first comments I made in the PR? I just edit that?

Yes, you can edit it.

chongchonghe commented:
> > > Can you update the PR description?
> > 
> > 
> > I'll do it this noon when I get to work. What is exactly a PR description? Is it the first comments I made in the PR? I just edit that?
> 
> Yes, you can edit it.

Done.

BenWibking reviewed (approved):
Looks good to me!
